### PR TITLE
chore: adding db.row_count for caching integrations

### PR DESCRIPTION
**Pull request by wconti27**
Adds `db.row_count` to all caching integrations to match new Datadog Span Spec. Only added for `get` `gets` and `get many` operations. 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [ ] Change is maintainable (easy to change, telemetry, documentation).
- [ ] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [ ] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).

## Reviewer Checklist

- [ ] Title is accurate.
- [ ] No unnecessary changes are introduced.
- [ ] Description motivates each change.
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Testing strategy adequately addresses listed risk(s).
- [ ] Change is maintainable (easy to change, telemetry, documentation).
- [ ] Release note makes sense to a user of the library.

### CODE REVIEW

File: ddtrace/contrib/dogpile_cache/region.py [hunk 1]
```diff
@@ -38,4 +44,8 @@ def _wrap_get_create_multi(func, instance, args, kwargs):
         span.set_tag("keys", keys)
         span.set_tag("region", instance.name)
         span.set_tag("backend", instance.actual_backend.__class__.__name__)
-        return func(*args, **kwargs)
+        response = func(*args, **kwargs)
+        if isinstance(response, Iterable):
```
Comment (mabdinur) [inline ddtrace/contrib/dogpile_cache/region.py:48]: What are the possible return values for `func(...)`? Can we document a case where it's not an iterable

File: ddtrace/contrib/pylibmc/client.py [hunk 2]
```diff
@@ -110,7 +115,17 @@ def _trace_cmd(self, method_name, *args, **kwargs):
...
 
-            return method(*args, **kwargs)
+            if method_name == "get":
+                result = method(*args, **kwargs)
+                span.set_metric(db.ROWCOUNT, 1 if result else 0)
+                return result
+            elif method_name == "gets":
+                result = method(*args, **kwargs)
+                # returns a tuple object that may be (None, None)
+                span.set_metric(db.ROWCOUNT, 1 if result[0] else 0)
```
Comment (mabdinur) [inline ddtrace/contrib/pylibmc/client.py:125]: Is result guaranteed to have at least one element? If not this will throw an index error. Also should we count all the elements in the array?

File: ddtrace/contrib/django/patch.py [hunk 3]
```diff
@@ -6,6 +6,10 @@
 `django.apps.registry.Apps.populate` is patched to add instrumentation for any
 specific Django apps like Django Rest Framework (DRF).
 """
+try:
+    from collections.abc import Iterable
```
Comment (mabdinur) [inline ddtrace/contrib/django/patch.py:10]: We use `typing.Iterable` in a few integrations but `collections.Iterable` here. We should try to be consistent. We can probably move lines 9 to 12 to `ddtrace/internal/compat.py` and import Iterable from this module.

File: ddtrace/contrib/dogpile_cache/region.py (hunk 1 above)
Comment (wconti27) [reply ddtrace/contrib/dogpile_cache/region.py:48]: Sequence[Any] for the wrapped function according to their API [here](https://dogpilecache.sqlalchemy.org/en/latest/api.html#dogpile.cache.region.CacheRegion.get_or_create_multi).

File: ddtrace/contrib/pylibmc/client.py (hunk 2 above)
Comment (wconti27) [reply ddtrace/contrib/pylibmc/client.py:125]: Gets is only a single "row" command, and returns a tuple mapping of its value and an ID. See [API](https://sendapatch.se/projects/pylibmc/reference.html#pylibmc.Client.gets:~:text=%2C%20cas_id)-,%C2%B6,-Get%20key%20and) here. Should be guaranteed but I guess we could check for length first to prevent a possible index error to be safe.

File: ddtrace/contrib/django/patch.py (hunk 3 above)
Comment (wconti27) [reply ddtrace/contrib/django/patch.py:10]: Fixed

File: ddtrace/contrib/django/patch.py [hunk 4]
```diff
@@ -139,7 +150,19 @@ def traced_cache(django, pin, func, instance, args, kwargs):
...
+        result = func(*args, **kwargs)
+        if command_name == "get_many":
+            # get_many returns a map
+            if result and isinstance(result, Iterable):
+                span.set_metric(db.ROWCOUNT, len(result))
+        elif command_name == "get":
+            if result:
+                # special case for Django~3.0 that returns an empty Sentinel object as missing key
+                if hasattr(instance, "_missing_key") and result == instance._missing_key:
+                    print("equals empty object", result == instance._missing_key)
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/django/patch.py:162]: This was more for debugging purposes as if was a failing test and couldn't figure out why. I re-worked this if to set the tag to 0 if the result is the _missing_key object and removed the print statement.

File: ddtrace/contrib/django/patch.py [hunk 5]
```diff
@@ -139,7 +150,19 @@ def traced_cache(django, pin, func, instance, args, kwargs):
             keys = utils.quantize_key_values(args[0])
             span.set_tag_str("django.cache.key", keys)
 
-        return func(*args, **kwargs)
+        result = func(*args, **kwargs)
+        if command_name == "get_many":
+            # get_many returns a map
+            if result and isinstance(result, Iterable):
+                span.set_metric(db.ROWCOUNT, len(result))
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/django/patch.py:157]: Setting to 0 if its empty, removed the nested if by just adding the logic within the tag set.

File: ddtrace/contrib/django/patch.py [hunk 6]
```diff
@@ -139,7 +150,19 @@ def traced_cache(django, pin, func, instance, args, kwargs):
...
+        elif command_name == "get":
+            if result:
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/django/patch.py:159]: same thing here

File: ddtrace/contrib/django/patch.py [hunk 7]
```diff
@@ -128,8 +130,12 @@ def traced_cache(django, pin, func, instance, args, kwargs):
         # set component tag equal to name of integration
         span.set_tag_str("component", config.django.integration_name)
 
+        # last string within the resource name is the command, ie: get
+        command_name = func.__name__
+
```
Comment (mabdinur) [inline ddtrace/contrib/django/patch.py:135]:
```suggestion
```

File: ddtrace/contrib/django/patch.py [hunk 8]
```diff
@@ -128,8 +130,12 @@ def traced_cache(django, pin, func, instance, args, kwargs):
...
         # update the resource name and tag the cache backend
         span.resource = utils.resource_from_cache_prefix(func_name(func), instance)
+
```
Comment (mabdinur) [inline ddtrace/contrib/django/patch.py:138]: unnecessary whitespace
```suggestion
```

File: ddtrace/contrib/django/patch.py [hunk 9]
```diff
@@ -139,7 +145,17 @@ def traced_cache(django, pin, func, instance, args, kwargs):
             keys = utils.quantize_key_values(args[0])
             span.set_tag_str("django.cache.key", keys)
 
-        return func(*args, **kwargs)
+        result = func(*args, **kwargs)
```
Comment (mabdinur) [inline ddtrace/contrib/django/patch.py:148]: we should move this variable to where it's used
```suggestion
        result = func(*args, **kwargs)
        command_name = func.__name__
```

File: ddtrace/contrib/django/patch.py [hunk 10]
```diff
@@ -139,7 +145,17 @@ def traced_cache(django, pin, func, instance, args, kwargs):
...
+        if command_name == "get_many":
+            span.set_metric(db.ROWCOUNT, len(result) if result and isinstance(result, Iterable) else 0)
+        elif command_name == "get":
+            # if valid result and check for special case for Django~3.0 that returns an empty Sentinel object as
+            # missing key
+            if result and not (hasattr(instance, "_missing_key") and result == instance._missing_key):
```
Comment (mabdinur) [inline ddtrace/contrib/django/patch.py:154]: this might be easier read

```suggestion
            if result and result != getattr(instance, "_missing_key", None):
```

File: .github/workflows/build-and-publish-image.yml [hunk 11]
```diff
@@ -0,0 +1,41 @@
+name: Build and publish image
```
Comment (mabdinur) [inline .github/workflows/build-and-publish-image.yml:1]: This is from a bad merge

File: ddtrace/contrib/flask_cache/tracers.py [hunk 12]
```diff
@@ -151,7 +154,12 @@ def get_many(self, *args, **kwargs):
             with self.__trace("flask_cache.cmd") as span:
                 span.resource = _resource_from_cache_prefix("GET_MANY", self.config)
                 span.set_tag(COMMAND_KEY, list(args))
-                return super(TracedCache, self).get_many(*args, **kwargs)
+                result = super(TracedCache, self).get_many(*args, **kwargs)
+
+                # get many returns a list, with either the key value or None if it doesn't exist
+                span.set_metric(db.ROWCOUNT, sum(1 for val in result if val))
+
```
Comment (mabdinur) [inline ddtrace/contrib/flask_cache/tracers.py:161]: nit: ws
```suggestion
```

File: ddtrace/contrib/flask_cache/tracers.py [hunk 13]
```diff
@@ -151,7 +154,12 @@ def get_many(self, *args, **kwargs):
             with self.__trace("flask_cache.cmd") as span:
                 span.resource = _resource_from_cache_prefix("GET_MANY", self.config)
                 span.set_tag(COMMAND_KEY, list(args))
-                return super(TracedCache, self).get_many(*args, **kwargs)
+                result = super(TracedCache, self).get_many(*args, **kwargs)
+
```
Comment (mabdinur) [inline ddtrace/contrib/flask_cache/tracers.py:158]: nit: ws
```suggestion
```

**Review by emmettbutler — COMMENTED** [review]
@wconti27 cleaning up the diff such that it only includes changes related to adding row_count for caching integrations would make this pull request much easier to review. Let me know when you've done that and I'll take another look.

File: ddtrace/contrib/pylibmc/client.py [hunk 14]
```diff
@@ -121,7 +134,14 @@ def _trace_multi_cmd(self, method_name, *args, **kwargs):
             if span and pre:
                 span.set_tag_str(memcached.QUERY, "%s %s" % (method_name, pre))
 
-            return method(*args, **kwargs)
+            if method_name == "get_multi":
+                result = method(*args, **kwargs)
+
+                # returns mapping of key -> value if key exists, but does not include a missing key. Empty result = {}
+                span.set_metric(db.ROWCOUNT, len(result) if result and isinstance(result, Iterable) else 0)
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/pylibmc/client.py:142]: Safety doesn't hurt!

File: ddtrace/contrib/dogpile_cache/region.py (hunk 1 above)
Comment (wconti27) [reply ddtrace/contrib/dogpile_cache/region.py]: removed this check

File: ddtrace/contrib/django/patch.py (hunk 7 above)
Comment (wconti27) [reply ddtrace/contrib/django/patch.py]: fixed

File: ddtrace/contrib/django/patch.py (hunk 9 above)
Comment (wconti27) [reply ddtrace/contrib/django/patch.py:160]: fixed

File: ddtrace/contrib/flask_cache/tracers.py (hunk 13 above)
Comment (wconti27) [reply ddtrace/contrib/flask_cache/tracers.py]: fixed

File: ddtrace/contrib/pylibmc/client.py [hunk 15]
```diff
@@ -111,7 +113,18 @@ def _trace_cmd(self, method_name, *args, **kwargs):
...
-            return method(*args, **kwargs)
+            if method_name == "get":
+                result = method(*args, **kwargs)
+                span.set_metric(db.ROWCOUNT, 1 if result else 0)
+                return result
+            elif method_name == "gets":
+                result = method(*args, **kwargs)
+
+                # returns a tuple object that may be (None, None)
+                span.set_metric(db.ROWCOUNT, 1 if result and isinstance(result, Iterable) and result[0] else 0)
```
Comment (mabdinur) [inline ddtrace/contrib/pylibmc/client.py:124]:
```suggestion
                span.set_metric(db.ROWCOUNT, 1 if isinstance(result, Iterable) and len(result) > 0 and result[0] else 0)
```

File: ddtrace/ext/sql.py [hunk 16]
```diff
@@ -6,7 +6,6 @@
 
 # tags
 QUERY = "sql.query"  # the query text
-ROWS = "sql.rows"  # number of rows returned by a query
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/ext/sql.py:9]: This is already deprecated within 1.x. This is the 2nd of 4 PR's for db.row_count, the first of which was already merged and made this change

[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/ext/sql.py:9]: oh lol its because I forgot to add the constant back since the other PR didn't end up deleting it either, fixed

File: ddtrace/contrib/flask_cache/tracers.py [hunk 17]
```diff
@@ -151,7 +154,10 @@ def get_many(self, *args, **kwargs):
             with self.__trace("flask_cache.cmd") as span:
                 span.resource = _resource_from_cache_prefix("GET_MANY", self.config)
                 span.set_tag(COMMAND_KEY, list(args))
-                return super(TracedCache, self).get_many(*args, **kwargs)
+                result = super(TracedCache, self).get_many(*args, **kwargs)
+                # get many returns a list, with either the key value or None if it doesn't exist
+                span.set_metric(db.ROWCOUNT, sum(1 for val in result if val))
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddtrace/contrib/flask_cache/tracers.py:159]: did this. implemented sum everywhere so if any values are None within the list, they will not be counted

**Review by emmettbutler — APPROVED** [review]
